// eska-common/dist/byte_serializer/models/map/simpleArray.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const modelType_1 = require("./../modelType");
const HEADER_LENGTH = 4;
class SimpleArray extends modelType_1.ModelType {
    constructor(type) {
        super();
        this.length.head = HEADER_LENGTH;
        this.typeElement = type;
    }
    getLength(value) {
        return this.length.head + this.getBodyLength(value);
    }
    getBodyLength(value) {
        return value.reduce((size, item) => {
            return size + this.typeElement.getLength(item);
        }, 0);
    }
    read(buffer, offset) {
        const sizeBody = buffer.readUInt32BE(offset);
        offset += this.length.head;
        const res = [];
        const end = offset + sizeBody;
        while (offset < end) {
            const result = this.typeElement.read(buffer, offset);
            res.push(result.value);
            offset = result.offset;
        }
        return {
            value: res,
            offset: offset
        };
    }
    write(buffer, value, offset) {
        const sizeBody = this.getBodyLength(value);
        buffer.writeUInt32BE(sizeBody, offset);
        offset += this.length.head;
        value.forEach(element => {
            offset = this.typeElement.write(buffer, element, offset);
        });
        return offset;
    }
}
exports.SimpleArray = SimpleArray;
