// eska-common/dist/byte_serializer/models/utf8.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const modelType_1 = require("./modelType");
const types_1 = require("src/byte_serializer/types");
const HEADER_LENGTH = 2;
class Utf8 extends modelType_1.ModelType {
    constructor() {
        super();
        this.length.head = HEADER_LENGTH;
        this.typeId = types_1.BufferTypesId.Utf8;
    }
    getLength(value) {
        return this.length.type + this.length.head + Buffer.byteLength(value);
    }
    read(buffer, offset) {
        offset += this.length.type;
        const size = buffer.readUInt16BE(offset);
        offset += this.length.head;
        return {
            value: buffer.toString('utf8', offset, offset += size),
            offset: offset
        };
    }
    writeHead(buffer, bodyLength, offset) {
        return buffer.writeUInt16BE(bodyLength, offset);
    }
    write(buffer, value, offset) {
        this.length.body = Buffer.byteLength(value);
        offset = this.writeTypeId(buffer, offset);
        offset = this.writeHead(buffer, this.length.body, offset);
        buffer.write(value, offset);
        return offset + this.length.body;
    }
}
exports.Utf8 = Utf8;
