// eska-common/dist/byte_serializer/types.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const int8_1 = require("./models/int8");
const uint8_1 = require("./models/uint8");
const int16_1 = require("./models/int16");
const uint16_1 = require("./models/uint16");
const int32_1 = require("./models/int32");
const uint32_1 = require("./models/uint32");
const uint64_1 = require("./models/uint64");
const double_1 = require("./models/double");
const utf8_1 = require("./models/utf8");
const array_1 = require("./models/array");
const object_1 = require("./models/object");
const dynamicBuffer_1 = require("./models/dynamicBuffer");
const map_1 = require("./models/map");
const set_1 = require("./models/set");
const boolean_1 = require("./models/boolean");
const number64_1 = require("./models/number64");
exports.BufferTypes = {
    Int8: int8_1.Int8,
    Uint8: uint8_1.Uint8,
    Int16: int16_1.Int16,
    Uint16: uint16_1.Uint16,
    Int32: int32_1.Int32,
    Uint32: uint32_1.Uint32,
    Uint64: uint64_1.Uint64,
    Double: double_1.Double,
    Utf8: utf8_1.Utf8,
    Array: array_1.BufferArray,
    Object: object_1.BufferObject,
    Buffer: dynamicBuffer_1.DynamicBuffer,
    Set: set_1.BufferSet,
    Map: map_1.BufferMap,
    Boolean: boolean_1.Bool,
    Number64: number64_1.Number64,
};
var BufferTypesId;
(function (BufferTypesId) {
    BufferTypesId[BufferTypesId["Int8"] = 0] = "Int8";
    BufferTypesId[BufferTypesId["Uint8"] = 1] = "Uint8";
    BufferTypesId[BufferTypesId["Int16"] = 2] = "Int16";
    BufferTypesId[BufferTypesId["Uint16"] = 3] = "Uint16";
    BufferTypesId[BufferTypesId["Int32"] = 4] = "Int32";
    BufferTypesId[BufferTypesId["Uint32"] = 5] = "Uint32";
    BufferTypesId[BufferTypesId["Uint64"] = 6] = "Uint64";
    BufferTypesId[BufferTypesId["Double"] = 7] = "Double";
    BufferTypesId[BufferTypesId["Utf8"] = 8] = "Utf8";
    BufferTypesId[BufferTypesId["BufferArray"] = 9] = "BufferArray";
    BufferTypesId[BufferTypesId["BufferObject"] = 10] = "BufferObject";
    BufferTypesId[BufferTypesId["BufferSet"] = 11] = "BufferSet";
    BufferTypesId[BufferTypesId["BufferMap"] = 12] = "BufferMap";
    BufferTypesId[BufferTypesId["Bool"] = 13] = "Bool";
    BufferTypesId[BufferTypesId["Number64"] = 14] = "Number64";
})(BufferTypesId = exports.BufferTypesId || (exports.BufferTypesId = {}));
exports.mapPrimitiveCode = new Map();
exports.mapPrimitiveCode.set(BufferTypesId.Int8, int8_1.Int8);
exports.mapPrimitiveCode.set(BufferTypesId.Uint8, uint8_1.Uint8);
exports.mapPrimitiveCode.set(BufferTypesId.Int16, int16_1.Int16);
exports.mapPrimitiveCode.set(BufferTypesId.Uint16, uint16_1.Uint16);
exports.mapPrimitiveCode.set(BufferTypesId.Int32, int32_1.Int32);
exports.mapPrimitiveCode.set(BufferTypesId.Uint32, uint32_1.Uint32);
exports.mapPrimitiveCode.set(BufferTypesId.Uint64, uint64_1.Uint64);
exports.mapPrimitiveCode.set(BufferTypesId.Double, double_1.Double);
exports.mapPrimitiveCode.set(BufferTypesId.Utf8, utf8_1.Utf8);
exports.mapPrimitiveCode.set(BufferTypesId.BufferArray, array_1.BufferArray);
exports.mapPrimitiveCode.set(BufferTypesId.BufferObject, object_1.BufferObject);
exports.mapPrimitiveCode.set(BufferTypesId.BufferSet, set_1.BufferSet);
exports.mapPrimitiveCode.set(BufferTypesId.BufferMap, map_1.BufferMap);
exports.mapPrimitiveCode.set(BufferTypesId.Bool, boolean_1.Bool);
exports.mapPrimitiveCode.set(BufferTypesId.Number64, number64_1.Number64);
